homeassistant/loader.py

```diff
@@ -63,6 +63,72 @@ def manifest_from_legacy_module(domain: str, module: ModuleType) -> Dict:
     }
 
 
+async def _async_get_custom_components(
+        hass: 'HomeAssistant') -> Dict[str, 'Integration']:
+    """Return list of custom integrations."""
+    try:
+        import custom_components
+    except ImportError:
+        return {}
+
+    def get_sub_directories(paths: List) -> List:
+        """Return all sub directories in a set of paths."""
+        return [
+            entry
+            for path in paths
+            for entry in pathlib.Path(path).iterdir()
+            if entry.is_dir()
+        ]
+
+    dirs = await hass.async_add_executor_job(
+        get_sub_directories, custom_components.__path__)
+
+    integrations = await asyncio.gather(*[
+        hass.async_add_executor_job(
+            Integration.resolve_from_root,
+            hass,
+            custom_components,
+            comp.name)
+        for comp in dirs
+    ])
+
+    return {
+        integration.domain: integration
+        for integration in integrations
+        if integration is not None
+    }
+
+
+async def async_get_custom_components(
+        hass: 'HomeAssistant') -> Dict[str, 'Integration']:
+    """Return cached list of custom integrations."""
+    cache = hass.data.get(DATA_CUSTOM_COMPONENTS)
+    if cache is None:
+        cache = hass.data[DATA_CUSTOM_COMPONENTS] = {}
+        cache['lock'] = asyncio.Lock()
```

Comment:
Instead of a lock, let's use the approach we use for entity registry using an event https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity_registry.py#L309-L326

Reply:
since there now exist another place. maybe that function decorator make sense after all :)